src/camera/Camera.h

```diff
@@ -0,0 +1,295 @@
+	*/
+	Camera();
+	/**
+	   @brief Opens a Camera that is not already open.
+
+	   Will open the given file for input and use the given intrinsic and
+	   extrinsic parameters, similar to the constructor.
+
+	   @returns true if opening the camera succeeds, and false if not.
+	 */
+	bool open(std::string filename, CameraParams intrinsic_params = CameraParams(),
+			  cv::Mat extrinsic_params = cv::Mat());
+	/**
+	   @brief Opens a Camera that is not already open.
+
+	   Will open the camera at the given camera ID and use the given intrinsic
+	   and extrinsic parameters, similar to the constructor.
+
+	   @returns true if opening the camera succeeds, and false if not.
+	*/
+	bool open(int camera_id, CameraParams intrinsic_params = CameraParams(),
+			  cv::Mat extrinsic_params = cv::Mat());
+	/**
+	   @brief Constructs a Camera that will open the given file for input and
+	   have the given name and description.
+
+	   @param filename The file to open and read a video feed from. This may be
+	   the name of a video file, or a URI of a video stream. This will be passed
+	   to the underlying OpenCV VideoCapture object, so anything supported by
+	   VideoCapture is supported here.
+
+	   @param name The name of the camera. This should ideally be unique, but
+	   isn't enforced at this time.
+
+	   @param description An optional description of the camera.
+	 */
+	Camera(std::string filename, std::string name, std::string description = "",
+		   CameraParams intrinsic_params = CameraParams(),
+		   cv::Mat extrinsic_params = cv::Mat());
+	/**
+	   @brief Constructs a Camera that will open the camera with the given ID
+	   and have the given name and description.
+
+	   @param filename The file to open and read a video feed from. This may be
+	   the name of a video file, or a URI of a video stream. This will be passed
+	   to the underlying OpenCV VideoCapture object, so anything supported by
+	   VideoCapture is supported here.
+
+	   @param name The name of the camera. This should ideally be unique, but
+	   isn't enforced at this time.
+
+	   @param description An optional description of the camera.
+	*/
+	Camera(int camera_id, std::string name, std::string description = "",
+		   CameraParams intrinsic_params = CameraParams(),
+		   cv::Mat extrinsic_params = cv::Mat());
+	/**
+	   @brief Copy constructor.
+
+	   A Camera object created through this constructor will share access to the
+	   underlying camera with the Camera object it is copying. When all Camera
+	   objects sharing access are destroyed, the camera will be closed.
+	 */
+	Camera(const Camera &other);
+
+	/**
+	   @brief Opens the camera using the given configuration file.
+
+	   @param filename The path to the configuration file to open and
+	   read. Configuration file should be formatted as described in @ref
+	   cameraconfig.
+
+	   @throws invalid_camera_config If the configuration is invalid for any
+	   reason.
+	*/
+	bool openFromConfigFile(std::string filename);
```

Comment:
I wonder if this should just be another overloaded `open()` method? I.e. `bool open(std::string config_filename);`

Reply:
Yeah, perhaps. Not sure how to handle the conflict with `open(std::string filename)` though, since you can also open a camera with a filename or URI as a string, so I just made it a separate method. It was originally just a public function that lived in the `cam` namespace, but when I refactored the Camera class to add a default constructor I thought it would work better as a method of `Camera`.

Reply:
Oh I see, I didn't notice there was already an `open` method with a string argument. In that case maybe the best way is just to put those two functions next to each other in the header file / implementation file. Not super important though, obviously